RAFT: Implement persistent WAL. Handle node restarts. Batch network messages.

manishrjain opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/289)
<!-- Reviewable:end -->

pawanrawal commented:
Quite a simple implementation of the WAL. 
A bit unrelated but we can get rid of the code in cluster package too.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[raftwal/wal.go, line 85 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM7p9nxVn-0aq8oH16:-KUM7p9nxVn-0aq8oH17:bhtwxrp) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/raftwal/wal.go#L85)):*
> ```Go
> 	}
> 
> 	if t > 0 || i > 0 {
> ```

Maybe we can add some comments as to why we do this? 

---

*[worker/draft.go, line 204 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM9FwlxMtVoCusTkWr:-KUM9FwmN3Z0yoiAhUc9:bv0vmao) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L204)):*
> ```Go
> 	case <-ctx.Done():
> 		return
> 	case <-ch:
> ```

Not sure about the utility of the error channel here.

---

*[worker/draft.go, line 332 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM7a2qQVV7Mlzp6Bg4:-KUM7a2qQVV7Mlzp6Bg5:b-utib5e) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L332)):*
> ```Go
> 	for {
> 		select {
> 		case t := <-time.Tick(time.Second):
> ```

Was this for debugging?

---

*[worker/draft.go, line 487 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6blBFpajRoS8aQaJ:-KUM6blBFpajRoS8aQaK:b8v9qua) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L487)):*
> ```Go
> 		restart = true
> 		if rerr = n.store.ApplySnapshot(sp); rerr != nil {
> 			return
> ```

Return error?

---

*[worker/draft.go, line 505 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6qK9zrlEJ86BemGo:-KUM6qK9zrlEJ86BemGp:b-cujpoa) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L505)):*
> ```Go
> 
> 	var term, idx uint64
> 	if !raft.IsEmptySnap(sp) {
> ```

Can be clubbed with the check above?

---

*[worker/draft.go, line 615 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUMAKK91Cz8-8I2RvlZ:-KUMAKK91Cz8-8I2Rvl_:by8ouf5) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L615)):*
> ```Go
> 	}
> 
> 	count := 0
> ```

Possibly we can get rid of this?

---

*[worker/groups.go, line 44 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6YwutMc3tfjmB-PT:-KUM6YwutMc3tfjmB-PU:bs7qsxm) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/groups.go#L44)):*
> ```Go
> 
> func StartRaftNodes(wal *raftwal.Wal, raftId uint64, my, peer string) {
> 	// ch := make(chan error, 2)
> ```

Is this for error handling?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-:-KUMBTJg2ErhFmi3yz4P:bhdzyey)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 7 of 11 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[raftwal/wal.go, line 85 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM7p9nxVn-0aq8oH16:-KUQifnmMH-t41MXYT7w:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/raftwal/wal.go#L85)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Maybe we can add some comments as to why we do this? 
</details>
Done.

---

*[worker/draft.go, line 204 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM9FwlxMtVoCusTkWr:-KUQiyHU2wezeoJ272zs:bcq2ou) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L204)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Not sure about the utility of the error channel here.
</details>
Done. Added a comment. But, basically we don't need to do anything if we get an error while sending message.

---

*[worker/draft.go, line 332 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM7a2qQVV7Mlzp6Bg4:-KUQj3TpGVt-5H_5_0sU:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L332)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Was this for debugging?
</details>
Done.

---

*[worker/draft.go, line 487 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6blBFpajRoS8aQaJ:-KUQj8CuBJwHjjlR6Dv4:b-xmljlm) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L487)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Return error?
</details>
It would. A return would return the then current values of restart and rerr.

---

*[worker/draft.go, line 505 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6qK9zrlEJ86BemGo:-KUQjYCmuN8j-ERwgL2P:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L505)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Can be clubbed with the check above?
</details>
Done.

---

*[worker/draft.go, line 615 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUMAKK91Cz8-8I2RvlZ:-KUQjaC6bi1CeIAh69o-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L615)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Possibly we can get rid of this?
</details>
Done.

---

*[worker/groups.go, line 44 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6YwutMc3tfjmB-PT:-KUQji1wAc4IIB0p3OGN:b-wnfmk9) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/groups.go#L44)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Is this for error handling?
</details>
Done. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/289)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[worker/draft.go, line 487 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-KUM6blBFpajRoS8aQaJ:-KUQmFqzDOQ_yj3meeML:b-wuvuws) ([raw file](https://github.com/dgraph-io/dgraph/blob/4a63da2b877928889d35e136f39f9bae236caba6/worker/draft.go#L487)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> It would. A return would return the then current values of restart and rerr.
</details>
Oh, right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/289#-:-KUQmQaE8jaWi2Webk4g:bnfp4nl)*
<!-- Sent from Reviewable.io -->
